Use logging module throughout

RichWellum opened the pull request:
Move from my verboseprint() to using logging module. The aim is also that when not running with --verbose there is very little output unless something fails.

None verbose output looks like:

```
rwellum@RWELLUM-M-34DF:[~/Desktop/Boxes]: iosxrv-x64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso
INFO:__main__:Creating Vagrant VirtualBox
INFO:__main__:Bringing up with iosxr_pexpect.py to install to disk and configure
INFO:__main__:Running basic unit tests on VirtualBox...
INFO:__main__:Passed basic test, box /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box is sane
```

Verbose looks like: (this is not all of it)

...Starting up

```
rwellum@RWELLUM-M-34DF:[~/Desktop/Boxes]: iosxrv-x64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso -v
DEBUG:__main__:Input ISO is iosxrv-fullk9-x64.iso
DEBUG:__main__:iosxrv-fullk9-x64.iso is a full image, RAM allocated is 4096 MB
INFO:__main__:Creating Vagrant VirtualBox
DEBUG:__main__:Virtual Box Manager Version: 5.0.24r108355

DEBUG:__main__:pathname: /Users/rwellum/Desktop/Boxes/iosxrv-x64-vbox
DEBUG:__main__:VM Name:  iosxrv-fullk9-x64
DEBUG:__main__:base_dir: /Users/rwellum/Desktop/Boxes/machines
DEBUG:__main__:box_dir:  /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64
DEBUG:__main__:box_out:  /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box
DEBUG:__main__:Found and deleted previous /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box
DEBUG:__main__:'iosxrv-fullk9-x64' is not running, nothing to poweroff
DEBUG:__main__:'iosxrv-fullk9-x64' is not registered, nothing to unregister and delete
DEBUG:__main__:Removing stale SSH entries
DEBUG:__main__:Create VM
DEBUG:__main__:Register VM
```

.... Waiting for vm to come up

```
INFO:__main__:logging in
2016-07-14,11:14:18 (iosxr_pexpect.py    :252 ) DEBUG (xr, port socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Connect
DEBUG:__main__:(xr, port socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Connect
2016-07-14,11:14:18 (iosxr_pexpect.py    :230 ) DEBUG (xr): Spawn (socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d)
DEBUG:__main__:(xr): Spawn (socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d)
2016-07-14,11:14:18 (iosxr_pexpect.py    :322 ) DEBUG (xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Waiting for XR login prompt
DEBUG:__main__:(xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Waiting for XR login prompt
2016-07-14,11:14:23 (iosxr_pexpect.py    :654 ) DEBUG (xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Send <enter>
DEBUG:__main__:(xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Send <enter>
2016-07-14,11:14:28 (iosxr_pexpect.py    :654 ) DEBUG (xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Send <enter>
DEBUG:__main__:(xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Send <enter>
2016-07-14,11:14:34 (iosxr_pexpect.py    :654 ) DEBUG (xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Send <enter>
```

..... VM is up and installing

```
[  116.634205] 00:03: ttyS0 at I/O 0x3f8 (irq = 4, base_baud = 115200) is a 16550A
[  116.685055] 00:04: ttyS1 at I/O 0x2f8 (irq = 3, base_baud = 115200) is a 16550A
[  116.737323] Non-volatile memory driver v1.3
[  116.745204] Linux agpgart interface v0.103
[  116.760442] brd: module loaded
[  116.780167] loop: module loaded
```

......Configuring the VM

```
  Enter root-system username:
2016-07-14,11:19:46 (iosxr_pexpect.py    :376 ) DEBUG (xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Got 'Enter root-system username:'
DEBUG:__main__:(xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Got 'Enter root-system username:'
2016-07-14,11:19:46 (iosxr_pexpect.py    :654 ) DEBUG (xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Send "vagrant"
DEBUG:__main__:(xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Send "vagrant"
2016-07-14,11:19:46 (iosxr_pexpect.py    :381 ) DEBUG (xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Sent XR login 'vagrant'
DEBUG:__main__:(xr: socat TCP:localhost:65000 -,raw,echo=0,escape=0x1d): Sent XR login 'vagrant'
2016-07-14,11:19:46 (iosxr_pexpect.py    :383 ) DEBUG Wait for password request
```

.....Finishing up and running ut

```
DEBUG:__main__:Waiting for machine to shutdown
DEBUG:__main__:Successfully shut down
DEBUG:__main__:Remove serial uarts before exporting
DEBUG:__main__:Compact VDI
DEBUG:__main__:Creating Virtualbox
DEBUG:__main__:Created: /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box
INFO:__main__:Running basic unit tests on VirtualBox...
INFO:__main__:Passed basic test, box /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box is sane
DEBUG:__main__:Single node use:
DEBUG:__main__: vagrant init 'IOS XRv'
DEBUG:__main__: vagrant box add --name 'IOS XRv' /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box --force
DEBUG:__main__: vagrant up
DEBUG:__main__:Multinode use:
DEBUG:__main__: Copy './iosxrv-x64-vbox/vagrantfiles/simple-mixed-topo/Vagrantfile' to the directory running vagrant and do:
DEBUG:__main__: vagrant box add --name 'IOS XRv' /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box --force
DEBUG:__main__: vagrant up
DEBUG:__main__: Or: 'vagrant up rtr1', 'vagrant up rtr2'
DEBUG:__main__:Note that both the XR Console and the XR linux shell username and password is vagrant/vagrant
DEBUG:__main__:'iosxrv-fullk9-x64' is not running, nothing to poweroff
DEBUG:__main__:'iosxrv-fullk9-x64' is registered, unregistering and deleting
```

Commit:
Use logging module throughout

msherif1234 commented on iosxr_iso2vbox.py:
```diff
@@ -69,29 +69,54 @@
+    Execute a CLI command in the 'background' as process.
+    '''
+    argv = shlex.split(cmd_string)
+    if debug is True:
```
i learnt using is as bad pratice why not if debug == True:

glennmatthews replied:
better practice still would be simply `if debug:`
best practice is to replace this with `logger.debug()` 😄

RichWellum replied:
logger.debug it is.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -69,29 +69,54 @@
-    argv = command_string.split()
+    argv = cmd_string.split()
     if debug is True:
         print ('argv: %s' % argv)
```
Why not:

``` python
logger.debug('argv: %s', argv)
```

Seems strange that you're adding print statements as part of a PR about moving away from them. 😄

RichWellum replied:
Ack. Cleaned up a few more remaining prints in the wild.

msherif1234 commented on iosxr_iso2vbox.py:
```diff
@@ -69,29 +69,54 @@
+    '''
+    argv = shlex.split(cmd_string)
+    if debug is True:
+        print ('argv: %s' % argv)
```
logger.debug() here instead of print ?

RichWellum replied:
Ack.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -152,58 +177,57 @@ def main(argv):
-            for arg in args:
-                print(arg,)
+        # Display all messages
+        logging.basicConfig(level=logging.DEBUG)
```
You can let argparse do some of this work for you:

``` python
    parser.add_argument('-v', '--verbose',
        action='store_const', const=logging.DEBUG,
        default=logging.INFO, help='turn on verbose messages')
...
    # Set logging level to >DEBUG or >INFO based on user selection
    # error > warning > info > debug
    # setLevel(logging.INFO) then you'll see error/warning/info messages but
    # not debug messages
    logging.basicConfig(level=args.verbose)
```

RichWellum replied:
That's much better - thanks!

RichWellum replied:
Propagated this throughout the tools.

msherif1234 commented on iosxr_iso2vbox.py:
```diff
@@ -69,29 +69,54 @@
+    argv = cmd_string.split()
+    if debug is True:
+        print ('argv: %s' % argv)
+    FNULL = open(os.devnull, 'w')
```
do u need to use with open() here as well so u don't need to worry about close() ?

glennmatthews replied:
Yes, agreed.

``` python
with open(os.devnull, 'w') as FNULL:
    return subprocess.call(argv, stdout=FNULL, stderr=FNULL)
```

goblinhack replied:
stderr to null ? bad...

RichWellum replied:
Combining both comments I get:
    with open(os.devnull, 'w') as FNULL:
        return subprocess.call(argv, stdout=FNULL, stderr=subprocess.STDOUT)

Seems ok?

goblinhack replied:
still don't like the string split. Does it work with

cat "justin bieber"

?

On 7/14/16, 3:25 PM, Richard Wellum wrote:

> > ```
> >   if debug is True:
> >       print ('argv: %s' % argv)
> >   result = subprocess.check_output(argv)
> >   return result
> > ```
> > 
> > +def run(cmd_string, debug=False):
> > -    '''
> > -    Execute a CLI command string quietly and return the result.
> > -    Note this does not handle operators which need whitespace.
> > -    In those cases the raw commands will need to be used.
> > -    E.g. passing grep 'a string separated by spaces'.
> > -    This is the bash equivalent of command > /dev/null.
> > -    '''
> > -    argv = cmd_string.split()
> > -    if debug is True:
> > -        print ('argv: %s' % argv)
> > -    FNULL = open(os.devnull, 'w')
> 
> Combining both comments I get:
>      with open(os.devnull, 'w') as FNULL:
>          return subprocess.call(argv, stdout=FNULL, stderr=subprocess.STDOUT)
> 
> Seems ok?
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70867444

RichWellum replied:
I call that out:

Execute a CLI command string quietly and return the result.
Note this does not handle operators which need whitespace.

glennmatthews replied:
Would be nice to have it take a list as input rather than a string so you don't have to do the split and "guess" what the user really wants. You could then do:

``` python
def run(args):
...

run('ssh-keygen -R [localhost]:2222'.split())
```

but frankly I'd be just as happy with:

``` python
run(['ssh-keygen', '-R', '[localhost]:2222'])
```

as it avoids any possible confusion about whitespace.

goblinhack replied:
That's a bit like saying this code is bugged and I know that 8)

On 7/14/16, 3:36 PM, Richard Wellum wrote:

> > ```
> >   if debug is True:
> >       print ('argv: %s' % argv)
> >   result = subprocess.check_output(argv)
> >   return result
> > ```
> > 
> > +def run(cmd_string, debug=False):
> > -    '''
> > -    Execute a CLI command string quietly and return the result.
> > -    Note this does not handle operators which need whitespace.
> > -    In those cases the raw commands will need to be used.
> > -    E.g. passing grep 'a string separated by spaces'.
> > -    This is the bash equivalent of command > /dev/null.
> > -    '''
> > -    argv = cmd_string.split()
> > -    if debug is True:
> > -        print ('argv: %s' % argv)
> > -    FNULL = open(os.devnull, 'w')
> 
> I call that out:
> 
> Execute a CLI command string quietly and return the result.
> Note this does not handle operators which need whitespace.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70869150

RichWellum replied:
FYI I am replacing all these run variations with one command. Won't string split. Errors will be seen in both verbose and non-verbose.

msherif1234 commented on iosxr_pexpect.py:
```diff
@@ -773,6 +764,13 @@ def main(self):
+            logging.basicConfig(level=logging.DEBUG)
+        else:
+            # Display info, warnings and errors
+            logging.basicConfig(level=logging.DEBUG)
```
did u mean logging.info here ?

RichWellum replied:
Ack.

Commit:
Make logging format universal and show function

Part 1

goblinhack commented on iosxr_iso2vbox.py:
```diff
@@ -69,29 +69,54 @@
+    if debug is True:
+        print ('argv: %s' % argv)
+    FNULL = open(os.devnull, 'w')
+    return(subprocess.call(argv, stdout=FNULL, stderr=subprocess.STDOUT))
```
You don't need argv split. You can do this: 

```
#
# Run a group of commands
#
def do_multi(self, command):
    self.logger.info('Run: %s', command)
    process = subprocess.Popen(command,stdout=subprocess.PIPE, shell=True)
    proc_stdout = process.communicate()[0].strip()
    print(proc_stdout)
```

RichWellum replied:
I'm trying to avoid shell=True though as it spawns a shell.

glennmatthews replied:
Thinking about this more, I'm not a fan of sending stdout and stderr to /dev/null since that means if the command fails, we have no context as to why it failed. Let's discuss offline the scenario you're trying to implement here.

goblinhack replied:
log file would be better for both

stderr should be logfile and output

On 7/14/16, 3:51 PM, Glenn Matthews wrote:

> > ```
> >   if debug is True:
> >       print ('argv: %s' % argv)
> >   result = subprocess.check_output(argv)
> >   return result
> > ```
> > 
> > +def run(cmd_string, debug=False):
> > -    '''
> > -    Execute a CLI command string quietly and return the result.
> > -    Note this does not handle operators which need whitespace.
> > -    In those cases the raw commands will need to be used.
> > -    E.g. passing grep 'a string separated by spaces'.
> > -    This is the bash equivalent of command > /dev/null.
> > -    '''
> > -    argv = cmd_string.split()
> > -    if debug is True:
> > -        print ('argv: %s' % argv)
> > -    FNULL = open(os.devnull, 'w')
> > -    return(subprocess.call(argv, stdout=FNULL, stderr=subprocess.STDOUT))
> 
> Thinking about this more, I'm not a fan of sending stdout and stderr to /dev/null since that means if the command fails, we have no context as to why it failed. Let's discuss offline the scenario you're trying to implement here.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70871517

RichWellum replied:
FYI I am replacing all these run variations with one command. Won't string split. Errors will be seen in both verbose and non-verbose.

goblinhack commented on iosxr_iso2vbox.py:
```diff
@@ -124,10 +149,10 @@ def main(argv):
     if re.search(':/', args.ISO_FILE):
         # URI Image
-        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
+        cmd_string = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
```
BTW why are you not doing

"{0}".format(bla)

style ?

RichWellum replied:
I need a little more context or an example? What I've done is common and flake8 is happy.

goblinhack replied:
'scp {0}@{1} .'.format(getpass.getuser(), args.ISO_FILE)

from what I've seen is the preferred way to do it now, not %

On 7/14/16, 3:26 PM, Richard Wellum wrote:

> > @@ -124,10 +149,10 @@ def main(argv):
> >       # Handle Input ISO (Local or URI)
> >       if re.search(':/', args.ISO_FILE):
> >           # URI Image
> > -        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> > -        cmd_string = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> 
> I need a little more context or an example? What I've done is common and flake8 is happy.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70867726

RichWellum replied:
@glennmatthews @msherif1234 Guys any preference here?

glennmatthews replied:
Either form is valid but `.format` is a bit more powerful.

I learned recently that you _should_ use %s-style for construction of logging messages because it permits lazy evaluation. For string construction, I tend to prefer .format but I don't have a strong opinion.

RichWellum replied:
.format is less readable to me. But maybe .format handles off characters in the scp string better.

I'll look into this - probably open a us.

goblinhack replied:
http://stackoverflow.com/questions/5082452/python-string-formatting-vs-format

was where I looked before and decided to change

fyi

On 7/14/16, 3:38 PM, Richard Wellum wrote:

> > @@ -124,10 +149,10 @@ def main(argv):
> >       # Handle Input ISO (Local or URI)
> >       if re.search(':/', args.ISO_FILE):
> >           # URI Image
> > -        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> > -        cmd_string = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> 
> @glennmatthews @msherif1234 Guys any preference here?
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70869399

goblinhack replied:
got an example of the should-use case ?

wish these emails had context... 8( see sawing between email the the browser

On 7/14/16, 3:42 PM, Glenn Matthews wrote:

> > @@ -124,10 +149,10 @@ def main(argv):
> >       # Handle Input ISO (Local or URI)
> >       if re.search(':/', args.ISO_FILE):
> >           # URI Image
> > -        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> > -        cmd_string = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> 
> Either form is valid but `.format` is a bit more powerful.
> 
> I learned recently that you _should_ use %s-style for construction of logging messages because it permits lazy evaluation. For string construction, I tend to prefer .format but I don't have a strong opinion.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70870010

goblinhack replied:
I don't care that much - it was just fyi

On 7/14/16, 3:45 PM, Richard Wellum wrote:

> > @@ -124,10 +149,10 @@ def main(argv):
> >       # Handle Input ISO (Local or URI)
> >       if re.search(':/', args.ISO_FILE):
> >           # URI Image
> > -        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> > -        cmd_string = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> 
> .format is less readable to me. But maybe .format handles off characters in the scp string better.
> 
> I'll look into this - probably open a us.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70870505

goblinhack replied:
I see:

Assuming you're using Python's logging module, you can pass the string formatting arguments as arguments to the .debug() method rather than doing the formatting yourself:

log.debug("some debug info: %s", some_info)

http://stackoverflow.com/questions/5082452/python-string-formatting-vs-format

so fair point on logging as an exception - if you care overly

neil

On 7/14/16, 3:50 PM, Neil McGill wrote:

> got an example of the should-use case ?
> 
> wish these emails had context... 8( see sawing between email the the browser
> 
> On 7/14/16, 3:42 PM, Glenn Matthews wrote:
> 
> > > @@ -124,10 +149,10 @@ def main(argv):
> > >       # Handle Input ISO (Local or URI)
> > >       if re.search(':/', args.ISO_FILE):
> > >           # URI Image
> > > -        cmdstring = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> > > -        cmd_string = 'scp %s@%s .' % (getpass.getuser(), args.ISO_FILE)
> > 
> > Either form is valid but `.format` is a bit more powerful.
> > 
> > I learned recently that you _should_ use %s-style for construction of logging messages because it permits lazy evaluation. For string construction, I tend to prefer .format but I don't have a strong opinion.
> > 
> > ---
> > 
> > You are receiving this because you commented.
> > Reply to this email directly or view it on GitHub:
> > https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70870010

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -69,29 +69,54 @@
-    Execute a CLI command string and return the result.
+    Execute a CLI command in the 'background' as process.
+    '''
+    argv = shlex.split(cmd_string)
```
I missed this before - why are you using `shlex.split(cmd_string)` here but `cmd_string.split()` below? Why the inconsistency?

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -229,44 +253,44 @@ def cleanup_vms(name, box_name):
     # Create and register a new VirtualBox VM
-    verboseprint('Create VM')
+    logger.debug('Create VM')
     run('VBoxManage createvm --name %s --ostype Linux26_64 --basefolder %s' % (vmname, base_dir))
```
just as a note - if you do as I suggested above and let your `run()` take a list of params, this would be IMHO clearer since it wouldn't need string substitution:

``` python
run(['VBoxManage', 'createvm', '--name', vm_name,
     '--ostype', 'Linux26_64', '--basefolder', base_dir])
```

RichWellum replied:
Doing this now.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -296,130 +320,140 @@ def cleanup_vms(name, box_name):
+    f.close()
+    rc = subprocess.call("chmod 766 ./runme.sh; ./runme.sh %s" % verbose_pipe, shell=True)
+    if rc == 0:
+        sys.exit('Configuring XR failed, exiting')
```
Yeah, this looks gross - thanks for pointing it out. :smile:

Your comment says it's being run in the background, but by using subprocess.call() you're blocking until it exits. By 'background' do you just mean 'quietly'?

This definitely looks overly convoluted to me. You're using Python to create a Bash script that run Python, then using Python to launch a shell that invokes that Bash script to run Python.... hmm.

goblinhack replied:
pybash-ception

On 7/14/16, 4:18 PM, Glenn Matthews wrote:

> > -    if verbose is True:
> > -        verbose_str = '-v'
> > -        verbose_pipe = ''
> > -    else:
> > -        # Shhhh...
> > -        verbose_str = ''
> > -        verbose_pipe = '>/dev/null 2>&1'
> >   +
> > -    cmd_string = "python %s %s -cmds 'socat TCP:localhost:%s -,raw,echo=0,escape=0x1d' -config iosxr_setup" % (verbose_str, iosxr_pexpect_path, console_port)
> > -    # Call socat connection from a script so it can be run in the background
> > -    f = open('runme.sh', 'w')
> > -    f.write(cmd_string)
> > -    f.close()
> > -    rc = subprocess.call("chmod 766 ./runme.sh; ./runme.sh %s" % verbose_pipe, shell=True)
> > -    if rc == 0:
> > -        sys.exit('Configuring XR failed, exiting')
> 
> Yeah, this looks gross - thanks for pointing it out. :smile:
> 
> Your comment says it's being run in the background, but by using subprocess.call() you're blocking until it exits. By 'background' do you just mean 'quietly'?
> 
> This definitely looks overly convoluted to me. You're using Python to create a Bash script that run Python, then using Python to launch a shell that invokes that Bash script to run Python.... hmm.
> 
> ---
> 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://github.com/ios-xr/iosxrv-x64-vbox/pull/10/files/fc29a814958e64ddff6a567916527f1c4084fe76#r70875862

RichWellum replied:
This is the convoluted nature of iso2vbox->pexpect->setup true. Yes by background I meant silently. Working on this.

glennmatthews commented on iosxr_pexpect.py:
```diff
@@ -781,11 +779,11 @@ def main(self):
         full_path_config_file = os.path.abspath(pathname) + '/' + self.opts.config + '.py'
 
         if self.opts.debug:
```
probably don't need this 'if' anymore?

RichWellum replied:
Ack.

glennmatthews commented on iosxr_store_box.py:
```diff
@@ -180,18 +172,19 @@ def verboseprint(*args):
-    print(email)
+    if args.verbose or test_only:
+        print('Email is:')
+        print(email)
```
Why not 'logger.debug()'?

RichWellum replied:
Because I have a --test option and I want to see the outputted email without sending it.

But it should be:     if args.verbose == logging.DEBUG or test_only:

glennmatthews commented on iosxr_test.py:
```diff
@@ -89,8 +92,8 @@ def bringup_vagrant():
-        print("pxssh failed on login")
-        print(str(e))
+        logger.debug("pxssh failed on login")
+        logger.debug(str(e))
```
Is it really the desired behavior to fail silently unless debugging is enabled?

RichWellum replied:
No these should all be logger.error - fixed them.

Commit:
First hunk of refactoring for logging.

Includes own pexpect handling to fascilitate the verbosity.

Lots of cleanup still required and some bugs.

Commit:
Further debugging and cleanups

Commit:
Updated readme

Added full help and non verbose output

Removed white space damage

RichWellum commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,295 @@
+        child.expect(prompt)
+
+        # Spin waiting for an ip address to be associated with the interface
+        while True:
```
This should be using xr_cli_wait_for_output() - not sure how I left that out but adding it back in and retesting.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -48,12 +48,10 @@
-Starts the VM, then uses iosxr_pexpect.py and iosxr_setup.py to configure
-XR and XR Aux for basic networking and XR Linux usage, with user name
-vagrant/vagrant.
+Starts the VM, then use pexpect to configure XR and XR Aux for
```
s/use/uses

RichWellum replied:
Ack.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,295 @@
+    '''
+    Set basic logging format.
+    '''
+    FORMAT = "[%(filename)s:%(lineno)s - %(funcName)20s() ] %(message)s"
```
whitespace between `()` and `]` intentional here?
Can you provide some examples of log output with this format?

RichWellum replied:
It put a little gap between the function name and the closing paren - e.g:

rwellum@RWELLUM-M-34DF:[~/Desktop/Boxes]: iosxrv-x64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso
[iosxr_iso2vbox.py:430 -                 main() ] Creating Vagrant VirtualBox
[iosxr_iso2vbox.py:155 -         configure_xr() ] Logging into Vagrant Virtualbox and configuring IOS XR
[iosxr_iso2vbox.py:586 -                 main() ] Powering down and generating Vagrant VirtualBox
[iosxr_iso2vbox.py:616 -                 main() ] Created: /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box
[iosxr_iso2vbox.py:626 -                 main() ] Running basic unit tests on Vagrant VirtualBox...
[iosxr_iso2vbox.py:647 -                 main() ] Passed basic test, box /Users/rwellum/Desktop/Boxes/machines/iosxrv-fullk9-x64/iosxrv-fullk9-x64.box is sane

But I think it's just as clear to remove that gap.

Example of the verbose - without the gap:

rwellum@RWELLUM-M-34DF:[~/Desktop/Boxes]: iosxrv-x64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso -v
[iosxr_iso2vbox.py:418 -                 main()] Input ISO is iosxrv-fullk9-x64.iso
[iosxr_iso2vbox.py:426 -                 main()] iosxrv-fullk9-x64.iso is a full image, RAM allocated is 4096 MB
[iosxr_iso2vbox.py:430 -                 main()] Creating Vagrant VirtualBox
[iosxr_iso2vbox.py:101 -                  run()] command: 'VBoxManage -v'

[iosxr_iso2vbox.py:106 -                  run()] Command succeeded with code 0:
[iosxr_iso2vbox.py:108 -                  run()] Output for: VBoxManage -v
[iosxr_iso2vbox.py:109 -                  run()] 5.0.26r108824

[iosxr_iso2vbox.py:433 -                 main()] Virtual Box Manager Version: 5.0.26r108824

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,295 @@
+    Allow the ability to hide errors and also to continue on errors.
+    '''
+    output = subprocess.Popen(cmd,
+                              stdin=subprocess.PIPE,
```
Unless you're actually supporting piping input into the subprocess, I don't think you want to set stdin here.

RichWellum replied:
Ack.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,295 @@
+    tup_output = output.communicate()
+
+    s_cmd = ' '.join(cmd)
+    logger.debug("command: '%s'\n", s_cmd)
```
Might want to move this to above the Popen call.

RichWellum replied:
Ack.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,295 @@
+        # Determine if the image is a crypto/k9 image or not
+        # This will be used to determine whether to configure ssh or not
+        child.sendline("bash -c rpm -qa | grep k9sec")
+        output = child.expect('-k9sec')
```
This isn't super-ideal as it means that for a non-k9 image, this will stall until expect() times out. Better to sendline and expect the next prompt (which should return quickly whether k9 or non-k9 regardless), then regexp against the received output to see if `-k9sec` is present.

RichWellum replied:
Good point thanks. Something like:

 child.sendline("bash -c rpm -qa | grep mgbl")
        child.expect(prompt)
        output = child.before
        if 'mgbl' in output:
            mgbl = True
            logger.debug("MGBL package detected")
        else:
            mgbl = False
            logger.debug("MGBL package not detected")

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -142,68 +402,33 @@ def main(argv):
     if not os.path.exists(input_iso):
-        print('==>', input_iso, 'does not exist')
-        sys.exit()
+        sys.exit('==>', input_iso, 'does not exist')
```
sys.exit() only takes a single argument, so this should be:

``` python
sys.exit('==> %s does not exist' % input_iso)
```

RichWellum replied:
Ack.

glennmatthews commented on iosxr_store_box.py:
```diff
@@ -105,43 +110,24 @@ def main(argv):
     if not os.path.exists(input_box):
-        print('==>', input_box, 'does not exist')
-        sys.exit()
+        sys.exit('==>', input_box, 'does not exist')
```
Ditto here about sys.exit args

RichWellum replied:
Ack.

ricobb commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,295 @@
+    Cleanup and unregister (delete) our working box.
+    '''
+    # Power off VM if it is running
+    vms_list_running = run(['vboxmanage', 'list', 'runningvms'])
```
command vboxmanage should be changed to VBoxManage

RichWellum replied:
Ack.

ricobb commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,295 @@
 
-    Note this does not handle operators which need whitespace.
+    # Unregister and delete
+    vms_list = run(['vboxmanage', 'list', 'vms'])
```
same here

RichWellum replied:
Ack.

ricobb commented on iosxr_iso2vbox.py:
```diff
@@ -142,68 +402,33 @@ def main(argv):
-        Unregisters and deletes any box with the passed in box_name
-        '''
-        # Remove any running boxes with the same name
-        vms_list_running = str(run('vboxmanage list runningvms'))
```
same here

ricobb commented on iosxr_iso2vbox.py:
```diff
@@ -142,68 +402,33 @@ def main(argv):
-            verboseprint("'%s' is not running, nothing to poweroff" % name)
-
-        # Unregister and delete any boxes with the same name
-        vms_list = str(run('vboxmanage list vms'))
```
seems there are several more instances with that require changes, please revert all

RichWellum replied:
These two are in deleted code. But I get your point and I've changed three instances. Thanks!

Commit:
Code review comments.

Added help output to readme.rst

ricobb commented on iosxr_pexpect.py:
```diff
@@ -1,810 +0,0 @@
-        # Get an IP address from the given interface
-        #
-        def get_cisco_ip_address(self, nic, debug="Get IP address"):
-            import re
```
Is the import here required?

glennmatthews replied:
This is deleted code (red background, `-` at left)

glennmatthews commented on README.rst:
```diff
@@ -85,17 +85,60 @@ How to use this tool
 
+5. Example with verbosity off
+
+iosxrv-x64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso
```
Please fence this into a code block (`::`) so it gets (non-)formatted properly.

glennmatthews commented on iosxr_test.py:
```diff
@@ -99,14 +99,17 @@ def bringup_vagrant():
         s = pexpect.pxssh.pxssh()
-        s.login(hostname, username, password, terminal_type, linux_prompt, login_timeout, port)
+        s.login(hostname, username, password, terminal_type, linux_prompt, login_timeout, port, auto_prompt_reset=False)
+        logger.debug('Sucessfully brought up VM and logged in')
```
s/Sucessfully/Successfully/

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -293,136 +513,149 @@ def cleanup_vms(name, box_name):
+        else:
+            logger.info('Passed basic test, box %s is sane', box_out)
+
+    logger.debug('Single node use:')
```
Feels a little odd to me that these instructions are debug logs - should this be print or logger.info instead?

RichWellum replied:
Yeah I agree - probably info's.

Commit:
Code review changes.

Couple of minor bugs unearthed like the wait for an ip address failing
due to an extra quote.

Commit:
Remove '-i' from README.

Commit:
Accidently removed bringup stage in test.

Added socat to README

Commit:
Added a loger, removed one commented line

Commit:
Remove scp test case.

Was not working and paramiko is a bear to install and maintain.

Commit:
Introduced a small bug

glennmatthews commented on README.rst:
```diff
@@ -85,20 +85,64 @@ How to use this tool
+5. Example with verbosity off
+
+   ::
+      iosxrv-x64-vbox/iosxr_iso2vbox.py iosxrv-fullk9-x64.iso
```
I think you need a blank line after the `::` - this isn't rendering properly on GitHub (https://github.com/ios-xr/iosxrv-x64-vbox/blob/cfad89e45ef99d2b27b0589cc0e4e289185fefaa/README.rst)

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -565,15 +578,7 @@ def main(argv):
-        #   v.customize ["modifyvm", :id, "--uart1", "0x3F8", 4, "--uartmode1", 'tcpserver', 65005]
-        #   v.customize ["modifyvm", :id, "--uart2", "0x2F8", 3, "--uartmode2", 'tcpserver', 65006]
-        # end
+    pause_to_debug()
```
I'd rather not turn debug into a global - perhaps instead:

``` python
if args.debug:
    pause_to_debug() # no longer checks global `debug` since you wouldn't have called it if not desired
```

RichWellum replied:
At the time I was adding this debug in a bunch of places, so wanted to just do a one liner. But now - will change back and remove the global.

glennmatthews commented on iosxr_test.py:
```diff
@@ -325,7 +302,7 @@ def main():
     logger.debug('result_linux=%s, result_xr=%s' % (result_linux, result_xr))
 
-    if result_linux is False or result_xr is False:
+    if result_linux is not True or result_xr is not True:
```
Ew. :) Why not `if not (result_linux and result_xr):` instead?

Commit:
Further CRR comments addressed.

Commit:
Removing ==>

RichWellum commented on README.rst:
```diff
@@ -85,20 +85,65 @@ How to use this tool
 
-2. Install VirtualBox and Vagrant (see guide below).
+2. Install VirtualBox, Vagrant and socat (see guide below).
 3. Download the appropriate ISO file, e.g. ``iosxrv-fullk9-x64.iso``
```
@glennmatthews Ran this overnight - 10x manually and several times in Jenkins and successfully build and passed ut each time.

https://engci-jenkins-rtp.cisco.com/jenkins/job/team_appdevci/job/iosxrv64_iso2vbox/37/console

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -65,38 +63,299 @@
+            logger.debug("MGBL package not detected")
+
+        # Wait for a management interface to be available
+        xr_cli_wait_for_output('sh run | inc MgmtEth', 'interface MgmtEth')
```
Just a nit - 'sh run **interface** | inc MgmtEth' might be slightly more efficient.

glennmatthews commented on iosxr_iso2vbox.py:
```diff
@@ -293,136 +523,131 @@ def cleanup_vms(name, box_name):
-    verboseprint(" vagrant box add --name 'IOS XRv' %s --force" % box_out)
-    verboseprint(' vagrant up')
-    verboseprint(" Or: 'vagrant up rtr1', 'vagrant up rtr2'")
+    if args.skip_test is False:
```
nit: `if not args.skip_test:` would be more Pythonic.

glennmatthews commented on iosxr_test.py:
```diff
@@ -72,25 +76,30 @@ def bringup_vagrant():
+        logger.error(e)
+
+    logger.debug('Waiting 30 seconds...')
+    time.sleep(30)
```
Do I remember you saying you were going to remove this?

RichWellum replied:
I had an issue when I did. Removing it now things are much more stable and retesting.

glennmatthews commented:
A few minor nits but otherwise looks good - 👍 Thanks for seeing this through!

Commit:
Last code review comments
